core/src/main/kotlin/transfer/Downloader.kt

```diff
@@ -0,0 +1,101 @@
+package com.juul.mcumgr.transfer
+
+import com.juul.mcumgr.McuMgrResult
+import com.juul.mcumgr.getOrElse
+import com.juul.mcumgr.getOrThrow
+import com.juul.mcumgr.onSuccess
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.channelFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+import okio.Buffer
+
+private const val RETRIES = 5
+
+interface Downloader {
+
+    suspend fun read(offset: Int): McuMgrResult<Response>
+
+    data class Event(val data: ByteArray, val offset: Int, val length: Int)
+    data class Response(val data: ByteArray, val offset: Int, val length: Int?)
+}
+
+@Throws
+fun Downloader.flow(
+    capacity: Int
+) = channelFlow {
```

Comment:
I think I kind of understand what's going on here, but if so, then it doesn't seem you need a `Flow`.
`Flow` is useful when you want to provide a lazy on-demand stream of data, but from what I can tell, you'll explicitly want to perform the upload. i.e. Standard Coroutines will accomplish what you want and me a lower cognitive lift to understand the code.

Reply:
Sounds good. I think will implement this with a returned receive channel instead.

Reply:
Not sure returning a `Channel` is the right approach either. Either that or I'm very much misunderstanding the purpose of this function.

Don't you just want to call a function and perform some downloads?
Why does that need to be "subscription" based? Why can't it be a `suspend` function (or `CoroutineScope` extension function)?

Unless this is an ongoing data stream? Whereas downloads could "appear" at any given time, and the `Flow` allows a library consumer to subscribe and be informed of these sporadic download events?

Reply:
I'm not set on using a Flow by any means, however I'd like the transfer to be a stream for progress callbacks which can drive a UI element (e.g. image upload used in DFU).

Reply:
Oh gotchya. For the purposes of propagating status/state a data stream does make sense.

What about this for an API?:

```kotlin
class Downloader(private resource: URI) {
    val progress: Flow<Int> // hot conflated Flow that emits values from 0-100
    suspend fun download(): ByteArray
}
```

Here, a call to `download` will suspend until complete. If developer wants status of the download (i.e. for UI updating purposes) then they can optionally subscribe to `progress`.

You can definitely make `progress` more fancy, with a `sealed class` that represents things like `InProgress(percent: Float)`, `Complete` or `Failure`.

Subscribing or unsubscribing to `progress` has no effect on the download (so is fine to subscribe from, for example, a `LiveData` that may unsubscribe and re-subscribe based on lifecycle; i.e. device rotation).

You `download` function can throw an exception on failure, or return a domain-specific `Result` object; up to you.

Reply:
You can get ultra fancy with something like:

```kotlin
suspend fun download(
    resource: URI,
    progressHandler: ((Int) -> Unit)? = null
): ByteArray {
    val downloader = Downloader(resource)
    return coroutineScope { 
        val job = downloader.progress.onEach { progressHandler?.invoke(it) }.launchIn(this)
        val result = downloader.download()
        progressHandler?.invoke(100)
        job.cancel()
        result
    }
}
```

It's usage could like:

```kotlin
val bytes = download(uri) { progress -> /* todo: update UI */ }
```